File: js-template/leetcode/editor/cn/linked-list-cycle.js
/*
 * @lc app=leetcode.cn id=141 lang=javascript
 * @lcpr version=30404
 *
 * [141] 环形链表
 */


import {ListNode} from "../common/listNode.js";
import {TreeNode} from "../common/treeNode.js";

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function(head) {
    var set = new Set();
    var pivot = head;
    while (pivot) {
        if (set.has(pivot))
            return true;
        set.add(pivot);
        pivot = pivot.next;
    }
    return false;
};
// @lc code=end

// your test code here





/*
// @lcpr case=start
// [3,2,0,-4]\n1\n
// @lcpr case=end

// @lcpr case=start
// [1,2]\n0\n
// @lcpr case=end

// @lcpr case=start
// [1]\n-1\n
// @lcpr case=end

 */

File: cpp-template/leetcode/editor/cn/linked-list-cycle.cpp
/*
 * @lc app=leetcode.cn id=141 lang=cpp
 * @lcpr version=30404
 *
 * [141] 环形链表
 */

#include <iostream>
#include <vector>
#include <string>
#include "../common/ListNode.cpp"
#include "../common/TreeNode.cpp"

using namespace std;

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode *dummyNode = new ListNode(0, head);

        ListNode *slow = dummyNode;
        ListNode *fast = dummyNode;

        while (fast && fast->next)
        {
            fast = fast->next->next;
            slow = slow->next;
            if (fast == slow) {
                delete dummyNode;
                return true;
            }
        }

        delete dummyNode;
        return false;
    }
};
// @lc code=end

int main() {
    Solution solution;
    // your test code here
}



/*
// @lcpr case=start
// [3,2,0,-4]\n1\n
// @lcpr case=end

// @lcpr case=start
// [1,2]\n0\n
// @lcpr case=end

// @lcpr case=start
// [1]\n-1\n
// @lcpr case=end

 */

